setwd("~/Dropbox/HIV")
setwd("C:/Users/yuchaoj/Dropbox/HIV")

library(Seurat)
library(Signac)
library(EnsDb.Hsapiens.v86)
library(GenomeInfoDb)
library(dplyr)
library(ggplot2)

library(chromVAR)
library(JASPAR2020)
library(TFBSTools)
library(motifmatchr)
library(BSgenome.Hsapiens.UCSC.hg38)

library(glmnet)
library(pheatmap)
library(fields)
library(qvalue)
library(gplots)
library(patchwork)
library(olsrr)
library(pdftools)
library(ape)
library(dendextend)

# From HIV 4
load('processed_data/hiv_predict.rda')

p1 <- DimPlot(hiv, reduction = "umap.rna", group.by='group', label = TRUE, label.size = 4, repel = TRUE) + ggtitle("scRNA-seq group") 
p2 <- DimPlot(hiv, reduction = "umap.atac", group.by = "group", label = TRUE, label.size = 4, repel = TRUE) + ggtitle("scATAC-seq group")
p3 <- DimPlot(hiv, reduction = "umap.rna", group.by = "seurat_clusters", label = TRUE, label.size = 4, repel = TRUE) + ggtitle("scRNA-seq cluster")
p4 <- DimPlot(hiv, reduction = "umap.atac", group.by = "seurat_clusters", label = TRUE, label.size = 4, repel = TRUE) + ggtitle("scATAC-seq cluster")
p = p1 + p2 + p3 +p4 +plot_layout(ncol=2)&theme(plot.title = element_text(hjust = 0.5))
p
ggsave(p, file='output/figure1_umap.pdf', width=9, height=6)

p<-VlnPlot(hiv, features = 'pRNA.sqrt')
p
ggsave(p, file='output/figure2_pRNA_condition.pdf', width=5, height=4)

p <- FeaturePlot(hiv, features = 'pRNA.sqrt', reduction = "umap.rna")
ggsave(p, file='output/figure2_pRNA_umap.pdf', width=5, height=4)


p<-VlnPlot(hiv, features = 'pRNA.sqrt', group.by = 'seurat_clusters')
p
ggsave(p, file='output/figure2_pRNA_cluster_vlnplot.pdf', width=6, height=4)

pdf(file='output/figure2_pRNA_vs_nATAC.pdf', width=4, height=4)
smoothScatter(hiv_ATAC_total, sqrt(pRNA), cex=1, pch=16, nrpoints = 0,
              xlab='Total HIV ATAC reads', ylab='sqrt percentage of HIV RNA reads')
points(hiv_ATAC_total, sqrt(pRNA), pch=16, cex=0.5, col='gray')
lm.run=lm(sqrt(pRNA)~hiv_ATAC_total)
abline(lm.run, col='black', lty=2)
title(paste('coeff =', signif(summary(lm.run)$coefficients[2,1],3), 
            'pval =', signif(summary(lm.run)$coefficients[2,4],3)))
dev.off()

toplot=cbind(pRNA=sqrt(pRNA), nATAC=hiv_ATAC_total)
toplot=as.data.frame(toplot)
toplot$nATAC=as.factor(toplot$nATAC)
p <- ggplot(toplot, aes(x=nATAC, y=pRNA)) + 
  geom_boxplot()+ggtitle(paste('pval =', signif(summary(lm.run)$coefficients[2,4],3)))
p
ggsave(p, file='output/figure2_pRNA_vs_nATAC_boxplot.pdf', width=2, height=4)

# From HIV 1
hiv.genes=tail(rownames(hiv@assays$SCT@scale.data))
p=DotPlot(hiv, features=paste0('sct_', hiv.genes), group.by='group')
ggsave(p, file='output/figure2_HIV_gene_RNA_dotplot.pdf', width=7, height=3)


cov_plot <- CoveragePlot(
  object = hiv,
  region = "HIV-1-6000",
  annotation = FALSE,
  peaks = TRUE,
  group.by = 'seurat_clusters'
)
p=cov_plot
ggsave(p, file='output/figure2_HIV_cov_atac_cluster.pdf', width=5, height=5)

cov_plot <- CoveragePlot(
  object = hiv,
  region = "HIV-1-6000",
  annotation = FALSE,
  peaks = TRUE,
  group.by = 'group'
)
p=cov_plot
ggsave(p, file='output/cov_hiv_cond.pdf', width=6, height=3)


# HIV 5
library(cowplot)
source('DotPlot.peak.R')
DotPlot.peak(hiv, features='hiv_ATAC_total', group.by='group', scale = FALSE)
p=DotPlot.peak(hiv, features='hiv_ATAC_total', group.by='seurat_clusters', scale=FALSE)
ggsave(p, file='output/figure2_atac_dotplot_by_cluster.pdf', width=4, height=5)



# Pairwise plot of p-values from condition-specific analyses (generated by Yuriko)
pval.gene=readRDS('./github/derived_data/mat.transformed.pval.gene.rds')
pval.TF=readRDS('./github/derived_data/mat.transformed.pval.tf.rds')

panel.cor <- function(x, y, digits = 2, prefix = "", cex.cor=NULL, ...)
{ usr <- par("usr"); on.exit(par(usr))
par(usr = c(0, 1, 0, 1))
filter=!is.na(y) & !is.na(x) & !is.nan(x) & !is.nan(y) & !is.infinite(x) & !is.infinite(y)
r <- cor(x[filter], y[filter])
r.temp=abs(r)
txt <- format(c(r, 0.123456789), digits = digits)[1]
txt <- paste0(prefix, 'r = ', txt)
cex.cor <- 1
text(0.5, 0.5, txt, cex = cex.cor * r.temp*2)}

pdf(file='output/supp_figure6_gene_venn.pdf', width=8, height=8)
pairs(pval.gene,
  upper.panel = panel.cor, lower.panel = function(x,y){smoothScatter(x,y,add=T)},
  main=('Condition-specific gene-linkage analyses'))
dev.off()

pdf(file='output/supp_figure6_TF_venn.pdf', width=8, height=8)
pairs(pval.TF,
      upper.panel = panel.cor, lower.panel = function(x,y){smoothScatter(x,y,add=T)},
      main=('Condition-specific TF-linkage analyses'))
dev.off()

